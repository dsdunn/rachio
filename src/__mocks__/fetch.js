export const getZones = jest.fn(() => Promise.resolve([{'all': 'the zones'}]));

export const startAll = jest.fn(() => Promise.resolve([{'all': 'the zones'}]));

