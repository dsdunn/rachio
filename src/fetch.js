//get zones (device)
//start zone, stop zone
//start/stop mulitiple zones

import { apiKey } from './apiKey';

const url = 'https://api.rach.io/1/public/';
const options = {
  headers: {
    Authorization: `Bearer ${apiKey}`
  }
}

export const getZones = (id) => {
  return fetch(url + 'device/' + id, options)
  .then(response => response.json())
  .then(results => results.zones.map(zone => ({
    id: zone.id,
    name: zone.name,
    zoneNumber: zone.zoneNumber
  })));
}