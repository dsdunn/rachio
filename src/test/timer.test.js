import React from 'react';
import { shallow } from 'enzyme';
import  Timer  from '../components/Timer';
import { startZone } from '../fetch';

jest.mock('../fetch');

window.fetch = jest.fn(() => Promise.resolve({'all': 'good'}));

describe('Timer', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Timer allOn={false} allTimeLeft={0} startOrStopAll={jest.fn()}/>);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should update state when props change', () => {
    expect(wrapper.state('running')).toEqual(false);
    wrapper.setProps({allOn: true});
    expect(wrapper.state('running')).toEqual(true);
  })

  it('should set intervalId in state when countdown is called', () => {
    expect(wrapper.state('intervalId')).toEqual('');
    wrapper.instance().countDown();
    expect(wrapper.state('intervalId')).not.toEqual('');
  })

  it('should setInterval when countDown is called', () => {
    jest.useFakeTimers();
    wrapper.instance().countDown();
    expect(setInterval).toHaveBeenCalledTimes(1);
  })

  it('calls stopCountDown when remaining reaches 0', () => {
    let spy = jest.spyOn(wrapper.instance(), 'stopCountDown');

    wrapper.setState({remaining: 2});
    wrapper.instance().countDown();
    expect(spy).not.toBeCalled();
    jest.advanceTimersByTime(2000);
    expect(spy).toBeCalled();
  })

  it('calls mockEditTime when id is "master-timer"', () => {
    let mockEditTime = jest.fn();

    wrapper.setProps({id: 'master-timer', editTime: mockEditTime});
    wrapper.setState({remaining: 2});
    wrapper.instance().countDown();
    jest.advanceTimersByTime(2000);
    expect(mockEditTime).toBeCalled();
  })

  it('resets state on stopCountDown', () => {
    wrapper.setState({running: true});
    wrapper.instance().stopCountDown();
    expect(wrapper.state('running')).toEqual(false);
    expect(wrapper.state('remaining')).toEqual(0);
  })

  it('updates state when input value changes', () => {
    let spy = jest.spyOn(wrapper.instance(), 'handleChange');

    wrapper.find('.timer-set-time').simulate('change', {target: {value: '2'}})
    expect(wrapper.state('remaining')).toEqual(120);
  })

  it('calls mockEditTime with remaining seconds when id is master-timer and input value changes', () => {
    let mockEditTime = jest.fn();

    wrapper.setProps({editTime: mockEditTime, id: 'master-timer'});
    wrapper.instance().handleChange({target: {value: 1}});
    expect(mockEditTime).toBeCalledWith(60);
  })

  it('reverses value of running when startOrStop is called', () => {
    expect(wrapper.state('running')).toEqual(false);
    wrapper.instance().startOrStop();
    expect(wrapper.state('running')).toEqual(true);
  })

  it('calls stopCountDown from startOrStop when running is true', () => {
    let spy = jest.spyOn(wrapper.instance(), 'stopCountDown');

    wrapper.setState({running: true});
    wrapper.instance().startOrStop();
    expect(spy).toBeCalled();
  })

  it('calls handleMasterStartOrStop from startOrStop when id is "master-timer"',() => {
    let spy = jest.spyOn(wrapper.instance(), 'handleMasterStartOrStop');
    wrapper.setProps({id: 'master-timer'});
    wrapper.instance().startOrStop(true);
    expect(spy).toBeCalled();
  })

  it('calls stopCountDown from handleMasterStartOrStop when running is true', () => {
    let spy = jest.spyOn(wrapper.instance(), 'stopCountDown');

    wrapper.setState({running: true});
    wrapper.instance().handleMasterStartOrStop();
    expect(spy).toBeCalled();
  })

  it('calls startOrStopAll when handleMasterStartOrStop is called', () => {
    let mockStartOfStopAll = jest.fn();

    wrapper.setProps({startOrStopAll: mockStartOfStopAll});
    wrapper.instance().handleMasterStartOrStop();
    expect(mockStartOfStopAll).toBeCalled();
  })
})












