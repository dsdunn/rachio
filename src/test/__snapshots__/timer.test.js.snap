// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timer
    allOn={false}
    allTimeLeft={0}
    startOrStopAll={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="timer-start-button "
          onClick={[Function]}
        >
          Start
        </button>,
        <div
          className="timer-form"
        >
          <input
            className="timer-set-time timer-countdown-display"
            max="180"
            min="0"
            onChange={[Function]}
            type="number"
            value={0}
          />
        </div>,
        false,
      ],
      "className": "timer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Start",
          "className": "timer-start-button ",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Start",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <input
            className="timer-set-time timer-countdown-display"
            max="180"
            min="0"
            onChange={[Function]}
            type="number"
            value={0}
          />,
          "className": "timer-form",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "timer-set-time timer-countdown-display",
            "max": "180",
            "min": "0",
            "onChange": [Function],
            "type": "number",
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="timer-start-button "
            onClick={[Function]}
          >
            Start
          </button>,
          <div
            className="timer-form"
          >
            <input
              className="timer-set-time timer-countdown-display"
              max="180"
              min="0"
              onChange={[Function]}
              type="number"
              value={0}
            />
          </div>,
          false,
        ],
        "className": "timer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Start",
            "className": "timer-start-button ",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Start",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="timer-set-time timer-countdown-display"
              max="180"
              min="0"
              onChange={[Function]}
              type="number"
              value={0}
            />,
            "className": "timer-form",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "timer-set-time timer-countdown-display",
              "max": "180",
              "min": "0",
              "onChange": [Function],
              "type": "number",
              "value": 0,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
